src/codec/data_storage/encrypted_data_chunk.rs

```diff
@@ -0,0 +1,109 @@
+use elliptic_curve::rand_core::CryptoRngCore;
+use futures::{AsyncWrite, AsyncWriteExt};
+use rand::Rng;
+use winnow::binary::le_u64;
+use winnow::error::ErrMode;
+use winnow::token::take;
+use winnow::Parser;
+
+use super::{data_chunk::DataChunk, data_options::DataOptions};
+use crate::codec::crypto::AccessKey;
+use crate::codec::crypto::AuthenticationTag;
+use crate::codec::crypto::Nonce;
+use crate::codec::{Cid, ParserResult, Stream};
+
+pub struct EncryptedDataChunk {
+    nonce: Nonce,
+    data: Box<[u8]>,
+    authentication_tag: AuthenticationTag,
+}
+
+impl EncryptedDataChunk {
+    pub fn new(nonce: Nonce, data: Box<[u8]>, authentication_tag: AuthenticationTag) -> Self {
+        Self {
+            nonce,
+            data,
+            authentication_tag,
+        }
+    }
+
+    pub async fn encode<W: AsyncWrite + Unpin + Send>(
+        &self,
+        writer: &mut W,
+    ) -> std::io::Result<(usize, Cid)> {
+        let mut encoded = Vec::new();
+
+        self.nonce.encode(&mut encoded).await?;
+        encoded.write_all(&self.data).await?;
+        self.authentication_tag.encode(&mut encoded).await?;
+
+        let cid = crate::utils::calculate_cid(&encoded);
+
+        writer.write_all(&encoded).await?;
+        Ok((encoded.len(), cid))
+    }
+
+    pub fn decrypt<'a>(
+        &'a self,
+        options: &DataOptions,
+        access_key: &AccessKey,
+    ) -> Result<DataChunk, DataChunkError> {
+        let mut plaintext_data = Vec::from(self.data.clone());
+        if let Err(err) = access_key.decrypt_buffer(
+            self.nonce.clone(),
+            &[],
+            &mut plaintext_data,
+            self.authentication_tag.clone(),
+        ) {
+            tracing::error!("failed to decrypt chunk: {err}");
+            return Err(DataChunkError::DecryptError);
+        }
+
+        let data_length = match le_u64::<&[u8], ErrMode<winnow::error::ContextError>>
+            .parse_peek(plaintext_data.as_slice())
+        {
+            Ok((_, length)) => length,
+            Err(err) => {
+                tracing::error!("failed to read inner length: {err:?}");
+                return Err(DataChunkError::PlainTextLengthParseError);
+            }
+        };
+        let plaintext_data: Vec<u8> = plaintext_data
+            .drain(8..(data_length as usize + 8))
```

Comment:
This + 8 everywhere is because that length isn't being taken into account for the inner chunk size, we should really update it everywhere to just be calculated correctly (but that doesn't have to happen in this PR).